### PR TITLE
[release-1.16] update envoy + manual cherry pick of #4286

### PR DESCRIPTION
**Pull request by dhawton**
**What this PR does / why we need it**:

Manual cherry-pick of #4286 and #4360 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: source/extensions/filters/http/istio_stats/istio_stats.cc [hunk 1]
```diff
@@ -574,42 +516,68 @@ struct Config : public Logger::Loggable<Logger::Id::filter> {
...
-      if (metric_overrides_->drop_.contains(metric)) {
+  // RAII for stream context propagation.
+  struct StreamOverrides {
+    StreamOverrides(Config& parent, const StreamInfo::StreamInfo& info,
+                    const Http::RequestHeaderMap* request_headers = nullptr,
+                    const Http::ResponseHeaderMap* response_headers = nullptr,
+                    const Http::ResponseTrailerMap* response_trailers = nullptr)
+        : parent_(parent) {
+      if (parent_.metric_overrides_) {
+        parent_.metric_overrides_->initialize(
```
Comment (kyessenov) [inline source/extensions/filters/http/istio_stats/istio_stats.cc:527]: This is buggy, need 7424dcd0b7931cfc550776ed2824a50e61fba50c as well.

**Comment by kyessenov** [comment]
Thanks for doing this.

File: source/extensions/filters/http/istio_stats/istio_stats.cc (hunk 1 above)
Comment (dhawton) [reply source/extensions/filters/http/istio_stats/istio_stats.cc]: I think there's some other changes missing.. there were some conflicts cherry picking and tests are failing now.

Comment (dhawton) [reply source/extensions/filters/http/istio_stats/istio_stats.cc]: We did the same pushes :)

**Review by johnma14 — APPROVED** [review]

**Pull request merged**